### PR TITLE
[UFC-1187] Support for the auth token based authentication in the SignIn action

### PR DESCRIPTION
**Pull request by exec-astraea**
## Checklist

-   Contains breaking changes ❌
-   Compatible with: MX10.12
-   Works in Android ✅
-   Works in iOS ✅
-   Works in Tablet ✅
-   Did you update version and changelog? ✅
-   PR title properly formatted (`[XX-000]: description`)? ✅

## This PR contains

-   [ ] Bug fix
-   [X] Feature
-   [ ] Refactor
-   [ ] Documentation
-   [ ] Other (describe)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased]
 
+- We added support for the `mx.login2` client API method in the SignIn action. This method allows to authenticate using tokens and stay signed in until the token expires.
```
Comment (EnasAbdelrazek) [inline CHANGELOG.md]: It is good to mention that this client API allows you to enable or disable an authentication token generation using the useAuthToken argument in online and offline-first apps.

File: src/client/SignIn.ts [hunk 2]
```diff
@@ -16,7 +16,7 @@ import { Big } from "big.js";
  * @param {string} password - This field is required.
```
Comment (EnasAbdelrazek) [inline src/client/SignIn.ts:16]: Please add the new parameter here and update the method description.

File: src/client/SignIn.ts [hunk 3]
```diff
@@ -26,7 +26,7 @@ export async function SignIn(username?: string, password?: string): Promise<Big>
         const onSuccess = (): void => resolve(new Big(200));
         const onError = (error: { status: number }): void => resolve(new Big(error.status));
 
-        mx.login(username, password, onSuccess, onError as any);
+        mx.login2(username, password, Boolean(useAuthToken), onSuccess, onError as any);
```
Comment (EnasAbdelrazek) [inline src/client/SignIn.ts]: This means offline apps have to set this optional argument to work as previously using an authentication token.

File: src/client/SignIn.ts [hunk 4]
```diff
@@ -22,6 +23,15 @@ export async function SignIn(username?: string, password?: string, useAuthToken?
         return Promise.resolve(new Big(401));
     }
 
+    if (typeof useAuthToken === "undefined") {
+        return new Promise(resolve => {
+            const onSuccess = (): void => resolve(new Big(200));
+            const onError = (error: { status: number }): void => resolve(new Big(error.status));
```
Comment (EnasAbdelrazek) [inline src/client/SignIn.ts]: Could you avoid duplication by extracting these onSuccess and onError handlers?

**Review by EnasAbdelrazek — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (exec-astraea) [reply CHANGELOG.md]: I clarified this part and added a link to the docs

File: src/client/SignIn.ts (hunk 3 above)
Comment (exec-astraea) [reply src/client/SignIn.ts]: Added a fallback to the old login method

File: src/client/SignIn.ts (hunk 2 above)
Comment (exec-astraea) [reply src/client/SignIn.ts:16]: Done

File: src/client/SignIn.ts (hunk 4 above)
Comment (exec-astraea) [reply src/client/SignIn.ts]: Fixed

File: src/client/SignIn.ts [hunk 5]
```diff
@@ -14,9 +14,10 @@ import { Big } from "big.js";
  *
  * @param {string} username - This field is required.
  * @param {string} password - This field is required.
+ * @param {string} useAuthToken - This field is optional.
```
Comment (furkanarabaci) [inline src/client/SignIn.ts]: Shouldn't this be `@param {boolean} useAuthToken - This field is optional`

**Review by furkanarabaci — CHANGES_REQUESTED** [review]

File: src/client/SignIn.ts (hunk 5 above)
Comment (exec-astraea) [reply src/client/SignIn.ts]: Fixed 👍

**Review by furkanarabaci — APPROVED** [review]
LGTM!

**Comment by EnasAbdelrazek** [comment]
login2 should be added to the mx typings.

**Pull request closed without merging**